Add DSMC module (#4125)
ECP-WarpX/WarpX: Docs/source/theory/collisions.rst

Collisions
==========

WarpX includes several different models to capture collisional processes
including collisions between kinetic particles (Coulomb collisions, DSMC,

> nuclear fusion) as well as collisions between kinetic particles and a fluid

Reviewer: ```suggestion
nuclear fusion) as well as collisions between kinetic particles and a fixed (i.e. non-evolving)
```
The word `fluid` here might be confusing ; some might understand this as interacting with the hybrid fluid-kinetic solver, or the fluid species in WarpX.

background species (MCC, background stopping).

.. _theory-collisions-mcc:

Monte Carlo Collisions

> ----------------------

Reviewer: ```suggestion
Monte Carlo Collisions (MCC)
------------------------------
```

Reviewer: Actually, should this be "Background Monte Carlo Collisions" (background MCC) ?

                 &= \frac{2Mmu^2}{M + m + \sqrt{M^2+m^2+2\gamma mM}}\frac{1}{\gamma + 1}
       \end{aligned}

where :math:`u` is the speed of the particle as tracked in WarpX (i.e.
:math:`u = \gamma v` with :math:`v` the particle speed), while :math:`m` and
:math:`M` are the rest masses of the simulation and background species,
respectively. The Lorentz factor is defined in the usual way,
:math:`\gamma \def \sqrt{1 + u^2/c^2}`. Note that if :math:`\gamma\to1` the above
expression reduces to the classical equation
:math:`E_{coll} = \frac{1}{2}\frac{Mm}{M+m} u^2`. The collision cross-sections
for all scattering processes are evaluated at the energy as calculated above.

Once a particle is selected for a specific collision process, that process determines how the particle is scattered as outlined below.

.. _theory-collisions-dsmc:

Direct Simulation Monte Carlo

> -----------------------------

Reviewer: ```suggestion
Direct Simulation Monte Carlo (DSMC)
--------------------------------------
```

> (Under construction)

Reviewer: Could you add some text here (even if it is short) to explain the approach?

Scattering processes
--------------------

Charge exchange
^^^^^^^^^^^^^^^

This process can occur when an ion and neutral (of the same species) collide
and results in the exchange of an electron. The ion velocity is simply replaced
with the neutral velocity and vice-versa.

Elastic scattering
^^^^^^^^^^^^^^^^^^

> The ``elastic`` option uses hard-sphere scattering, with a differential

Reviewer: I am not sure I understand the term `hard-sphere` in this context. 
Should we simply say that the scattering is isotropic, i.e. has a differential cross-section that is independent of the angle.

cross section that is independent of angle.
This scattering process as well as the ones below that relate to it, are all

> performed in the center-of-momentum (COM) frame. Designating the COM velocity of

Reviewer: When we use the MCC framework to perform elastic scattering, this text seems in contradiction with the paragraph saying "The particle velocity is then boosted to the stationary frame of the neutral through a Galilean transformation."

Author: There is actually two separate boosts in the collision process. First, we assume a non-relativistic neutral velocity which allows us to use a Galilean transformation to get to the rest frame of the neutral. In this frame the relativistic collision energy is easy to calculate with the function `ParticleUtils::getCollisionEnergy()`. This calculation happens very often and therefore is good to have a minimum number of FLOPS. Particles that are then chosen for a collision (presumably a relatively small number), is then Lorentz transformed to the COM frame (this therefore allows relativistic simulation particles) where the scattering happens. Does that clear things up?

The process is also the same as for elastic scattering except the excitation energy cost is subtracted from the particle energy. This is done by reducing the velocity before a scattering angle is chosen.

Benchmarks
----------

> See the :ref:`MCC example <examples-mcc-turner>` for a bencmark of the MCC

Reviewer: ```suggestion
See the :ref:`MCC example <examples-mcc-turner>` for a benchmark of the MCC
```
